File: ganttproject/src/net/sourceforge/ganttproject/ChartTabContentPanel.java
/*
GanttProject is an opensource project management tool. License: GPL3
Copyright (C) 2011 Dmitry Barashev, GanttProject Team

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 3
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package net.sourceforge.ganttproject;

import biz.ganttproject.core.option.ChangeValueEvent;
import biz.ganttproject.core.option.ChangeValueListener;
import com.google.common.base.Preconditions;
import com.google.common.base.Supplier;
import net.sourceforge.ganttproject.chart.TimelineChart;
import net.sourceforge.ganttproject.chart.overview.NavigationPanel;
import net.sourceforge.ganttproject.chart.overview.ZoomingPanel;
import net.sourceforge.ganttproject.gui.GanttImagePanel;
import net.sourceforge.ganttproject.gui.UIFacade;
import net.sourceforge.ganttproject.language.GanttLanguage;

import javax.swing.*;
import javax.swing.border.Border;
import java.awt.*;
import java.awt.event.ComponentAdapter;
import java.awt.event.ComponentEvent;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.util.ArrayList;
import java.util.List;

abstract class ChartTabContentPanel {
  private final TimelineChart myChart;
  private JSplitPane mySplitPane;
  private final List<Component> myPanels = new ArrayList<>();
  private final UIFacade myUiFacade;
  private int myImageHeight;
  private Supplier<Integer> myHeaderHeight;
  private GanttImagePanel myImagePanel;

  ChartTabContentPanel(IGanttProject project, UIFacade workbenchFacade, TimelineChart chart) {
    NavigationPanel navigationPanel = new NavigationPanel(project, chart, workbenchFacade);
    ZoomingPanel zoomingPanel = new ZoomingPanel(workbenchFacade, chart);
    addChartPanel(zoomingPanel.getComponent());
    addChartPanel(navigationPanel.getComponent());
    myUiFacade = workbenchFacade;
    myChart = Preconditions.checkNotNull(chart);
    myUiFacade.getMainFrame().addWindowListener(new WindowAdapter() {
      @Override
      public void windowOpened(WindowEvent windowEvent) {
        updateTimelineHeight();
      }
    });
  }

  JComponent createContentComponent() {
    JPanel tabContentPanel = new JPanel(new BorderLayout());
    final JPanel left = new JPanel(new BorderLayout());
    final Box treeHeader = Box.createVerticalBox();
    final JComponent buttonPanel = (JComponent) createButtonPanel();
    JPanel buttonWrapper = new JPanel(new BorderLayout());
    buttonWrapper.add(buttonPanel, BorderLayout.WEST);
    //button.setAlignmentX(Component.LEFT_ALIGNMENT);
    treeHeader.add(buttonWrapper);

    int defaultScaledHeight = (int)(UIFacade.DEFAULT_LOGO.getIconHeight() * myUiFacade.getDpiOption().getValue() / (1f * UIFacade.DEFAULT_DPI));
    myImagePanel = new GanttImagePanel(myUiFacade.getLogo(), 300, defaultScaledHeight);
    myImageHeight = myImagePanel.getPreferredSize().height;
    JPanel imageWrapper = new JPanel(new BorderLayout());
    imageWrapper.add(myImagePanel, BorderLayout.WEST);
    //myImagePanel.setAlignmentX(Component.LEFT_ALIGNMENT);
    treeHeader.add(imageWrapper);

    left.add(treeHeader, BorderLayout.NORTH);

    left.add(getTreeComponent(), BorderLayout.CENTER);
    Dimension minSize = new Dimension(0, 0);
    left.setMinimumSize(minSize);

    JPanel right = new JPanel(new BorderLayout());
    final JComponent chartPanels = createChartPanels();
    right.add(chartPanels, BorderLayout.NORTH);
    right.setBackground(new Color(0.93f, 0.93f, 0.93f));
    right.add(getChartComponent(), BorderLayout.CENTER);
    right.setMinimumSize(minSize);

    mySplitPane = new JSplitPane(JSplitPane.HORIZONTAL_SPLIT);
    if (GanttLanguage.getInstance().getComponentOrientation() == ComponentOrientation.LEFT_TO_RIGHT) {
      mySplitPane.setLeftComponent(left);
      mySplitPane.setRightComponent(right);
      mySplitPane.applyComponentOrientation(ComponentOrientation.LEFT_TO_RIGHT);
      mySplitPane.setDividerLocation(Math.min(300, left.getPreferredSize().width));
    } else {
      mySplitPane.setRightComponent(left);
      mySplitPane.setLeftComponent(right);
      mySplitPane.setDividerLocation(
          Toolkit.getDefaultToolkit().getScreenSize().width - Math.min(300, left.getPreferredSize().width));
      mySplitPane.applyComponentOrientation(ComponentOrientation.RIGHT_TO_LEFT);
    }
    mySplitPane.setOneTouchExpandable(true);
    mySplitPane.resetToPreferredSizes();
    tabContentPanel.add(mySplitPane, BorderLayout.CENTER);

    ChangeValueListener changeValueListener = new ChangeValueListener() {
      @Override
      public void changeValue(ChangeValueEvent event) {
        if (myUiFacade.getDpiOption().getValue() < 96) {
          return;
        }
        SwingUtilities.invokeLater(new Runnable() {
          @Override
          public void run() {
            alignTopPanelHeights(buttonPanel, chartPanels);
            myImagePanel.setScale(myUiFacade.getDpiOption().getValue() / (1f * UIFacade.DEFAULT_DPI));
            myImageHeight = myImagePanel.getHeight();
            updateTimelineHeight();
          }
        });
      }
    };
    myUiFacade.getDpiOption().addChangeValueListener(changeValueListener, 2);
    return tabContentPanel;
  }

  private void alignTopPanelHeights(JComponent buttonPanel, JComponent chartPanels) {
    int maxHeight = Math.max(buttonPanel.getSize().height, chartPanels.getSize().height);
    if (buttonPanel.getHeight() < maxHeight) {
      //left.setBorder(BorderFactory.createEmptyBorder(maxHeight - buttonPanel.getHeight(), 0, 0, 0));
      int diff = maxHeight - buttonPanel.getHeight();
      Border emptyBorder = BorderFactory.createEmptyBorder((diff+1)/2, 0, diff/2, 0);
      buttonPanel.setBorder(emptyBorder);
    }
    if (chartPanels.getHeight() < maxHeight) {
      int diff = maxHeight - chartPanels.getHeight();
      //Border emptyBorder = BorderFactory.createEmptyBorder((diff+1)/2, 0, diff/2, 0);
      //chartPanels.setBorder(emptyBorder);
      chartPanels.remove(chartPanels.getComponent(chartPanels.getComponentCount() - 1));
      chartPanels.add(Box.createRigidArea(new Dimension(0, diff)));
    }
  }

  protected abstract Component getChartComponent();

  protected abstract Component getTreeComponent();

  protected abstract Component createButtonPanel();

  int getDividerLocation() {
    return mySplitPane.getDividerLocation();
  }

  void setDividerLocation(int location) {
    mySplitPane.setDividerLocation(location);
  }

  private JComponent createChartPanels() {
    Box panelsBox = Box.createHorizontalBox();
    for (Component panel : myPanels) {
      panelsBox.add(panel);
      panelsBox.add(Box.createHorizontalStrut(10));
    }
    return panelsBox;
  }

  void addChartPanel(Component panel) {
    myPanels.add(panel);
  }

  protected UIFacade getUiFacade() {
    return myUiFacade;
  }

  private void updateTimelineHeight() {
    int timelineHeight = myHeaderHeight.get() + myImageHeight;
    myChart.setTimelineHeight(timelineHeight);
  }

  void addTableResizeListeners(final Component tableContainer, final Component table) {
    myHeaderHeight = new Supplier<Integer>() {
      @Override
      public Integer get() {
        if (table.isShowing() && tableContainer.isShowing()) {
          Point tableLocation = table.getLocationOnScreen();
          Point containerLocation = tableContainer.getLocationOnScreen();
          return tableLocation.y - containerLocation.y;
        } else {
          return 0;
        }
      }
    };
    ComponentAdapter componentListener = new ComponentAdapter() {
      @Override
      public void componentShown(ComponentEvent componentEvent) {
        updateTimelineHeight();
      }

      @Override
      public void componentResized(ComponentEvent componentEvent) {
        updateTimelineHeight();
      }

      @Override
      public void componentMoved(ComponentEvent componentEvent) {
        updateTimelineHeight();
      }
    };
    tableContainer.addComponentListener(componentListener);
    table.addComponentListener(componentListener);
  }

  public void setActive(boolean active) {
    if (active) {
      getTreeComponent().requestFocus();
      updateTimelineHeight();
    }
  }
}

File: ganttproject/src/main/java/net/sourceforge/ganttproject/ChartTabContentPanel.kt
/*
GanttProject is an opensource project management tool. License: GPL3
Copyright (C) 2011 Dmitry Barashev, GanttProject Team

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 3
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package net.sourceforge.ganttproject

import biz.ganttproject.app.FXToolbarBuilder
import biz.ganttproject.app.applicationFont
import biz.ganttproject.app.getGlyphIcon
import com.google.common.base.Preconditions
import de.jensd.fx.glyphs.fontawesome.FontAwesomeIcon
import de.jensd.fx.glyphs.fontawesome.FontAwesomeIconView
import javafx.scene.control.MenuItem
import javafx.scene.control.Skin
import javafx.scene.control.SplitMenuButton
import javafx.scene.layout.StackPane
import javafx.scene.text.Text
import net.sourceforge.ganttproject.action.GPAction
import net.sourceforge.ganttproject.chart.TimelineChart
import net.sourceforge.ganttproject.chart.overview.NavigationPanel
import net.sourceforge.ganttproject.chart.overview.ZoomingPanel
import net.sourceforge.ganttproject.gui.GanttImagePanel
import net.sourceforge.ganttproject.gui.UIFacade
import net.sourceforge.ganttproject.language.GanttLanguage
import java.awt.Component
import javax.swing.JComponent
import javax.swing.SwingUtilities

internal abstract class ChartTabContentPanel(
    protected val project: IGanttProject, workbenchFacade: UIFacade, chart: TimelineChart) {

  private var zoomingPanel: ZoomingPanel = ZoomingPanel(workbenchFacade, chart)
  private var navigationPanel: NavigationPanel = NavigationPanel(project, chart, workbenchFacade)
  private val myChart: TimelineChart
  private val myPanels: MutableList<Component> = ArrayList()
  private val myUiFacade: UIFacade
  private var myImagePanel: GanttImagePanel? = null
  protected var headerHeight: () -> Int = { 0 }
  protected var imageHeight: () -> Int = { myImagePanel?.preferredSize?.height ?: 0}


  protected open fun buildDropdownActions(): List<GPAction> = emptyList()
  protected open fun buildToolbarActions(): List<GPAction> = emptyList()

  abstract val chartComponent: JComponent?
    get

  protected abstract fun getTreeComponent(): Component?
  protected abstract fun createButtonPanel(): Component?

  fun addChartPanel(panel: Component) {
    myPanels.add(panel)
  }

  protected fun getUiFacade(): UIFacade {
    return myUiFacade
  }

  protected fun updateTimelineHeight() {
    //val timelineHeight = toolbar.component.height /* + myImageHeight*/
    SwingUtilities.invokeLater {
      val timelineHeight = headerHeight() + imageHeight()
      myChart.setTimelineHeight(timelineHeight)
      myChart.reset();
    }
  }

  init {

    addChartPanel(createNavigationToolbarBuilder().withScene().build().component)

    myUiFacade = workbenchFacade
    myChart = Preconditions.checkNotNull(chart)
  //++
  //    myUiFacade.mainFrame.addWindowListener(object : WindowAdapter() {
//      override fun windowOpened(windowEvent: WindowEvent) {
//        updateTimelineHeight()
//      }
//    })
  }

  fun createNavigationToolbarBuilder() = FXToolbarBuilder().also {
    it.withApplicationFont(applicationFont)
    zoomingPanel.buildToolbar(it)
    it.addWhitespace()
    navigationPanel.buildToolbar(it)
  }.withClasses("toolbar-common", "toolbar-small", "toolbar-chart")

  fun buildDropdown(titleAction: GPAction?, actions: List<GPAction>) =
    MyMenuButton().apply {
      setup(titleAction, actions)
    }
}

class MyMenuButton : SplitMenuButton() {
  private lateinit var arrowIcon: Text
  override fun createDefaultSkin(): Skin<*> {
    return super.createDefaultSkin().also {
      arrowIcon.styleClass.add("first")
      ((children[1] as StackPane).children[0] as StackPane).children.let {
        it.add(arrowIcon)
        it.add(FontAwesomeIconView(FontAwesomeIcon.ARROW_CIRCLE_DOWN).also {it.styleClass.add("second")})
      }
    }
  }

  fun setup(titleAction: GPAction?, actions: List<GPAction>) {
    styleClass.add("btn-create-item")
    text = "Add"
    arrowIcon = titleAction?.getGlyphIcon() ?: FontAwesomeIconView(FontAwesomeIcon.PLUS)
    items.addAll(actions.map { action ->
      MenuItem(GanttLanguage.correctLabel(action.localizedName)).also { item ->
        item.setOnAction { SwingUtilities.invokeLater { action.actionPerformed(null) }}
      }
    })
    setOnAction {
      titleAction?.let { SwingUtilities.invokeLater { it.actionPerformed(null) }}
    }
  }

}
